### PR TITLE
feat: loader and game flow on front-end

### LINKED ISSUE #38: Create Game Component

### PR DESCRIPTION
**Pull request by vkbinfo**
This Implements the following changes:-

- Gave 3 seconds delay after both parties have joined the game.
- Implemented Loader, so the user does not have to stare at a blank screen.
- Handled sending of snake body color from the background

Resolves #38

### CODE REVIEW

**Comment by vkbinfo** [comment]
ezgif com-video-to-gif

File: client/src/api/socketService.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import io from 'socket.io-client';
+
+// const socketServerUrl = 'https://snake-backend-1551222774563.appspot.com';
+// const socket = io(socketServerUrl);
+const socket = io('http://localhost:4000').connect();
```
Comment (prakhar1912) [inline client/src/api/socketService.js]: Why not use environment here?

File: client/src/components/Home.js [hunk 2]
```diff
@@ -1,30 +1,98 @@
...
+    });
+  }
+
+  playGame() {
+    this.setState({ loading: true });
+    socket.emit('joinGame', 'nothing');
+    socket.on('assignPlayerId', (playerId) => {
+      // eslint-disable-next-line no-console
+      console.log('playerId assigned: ', playerId);
+      this.state.playerId = playerId;
```
Comment (prakhar1912) [inline client/src/components/Home.js]: No `setState`?

File: client/src/components/Home.js [hunk 3]
```diff
@@ -1,30 +1,98 @@
...
+    });
+  }
+
+
+  toBeRenderJSX() {
```
Comment (prakhar1912) [inline client/src/components/Home.js]: Think of a different function name, like `handleNoRedirect`

File: client/src/components/Home.js [hunk 4]
```diff
@@ -1,30 +1,98 @@
...
+          <Button className={classes.button} variant="contained" color="secondary" size="large">
+            Find a Friend
+          </Button>
+        </Grid>
+        <Leaderboard limit={5} />
+      </div>
+    );
+  }
+
+  renderRedirectToGame() {
```
Comment (prakhar1912) [inline client/src/components/Home.js]: Or maybe just `redirectToGame`?

File: client/src/components/common/loaderComponent/loaderComponent.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+      <p>{message}</p>
+    </div>
+  );
+};
+
+LoaderComponent.propTypes = {
+  height: PropTypes.string.isRequired,
+  width: PropTypes.string.isRequired,
+  // eslint-disable-next-line react/require-default-props
+  message: PropTypes.string,
```
Comment (niinpatel) [inline client/src/components/common/loaderComponent/loaderComponent.js]: shouldn't this be required?

File: client/src/components/gameBoardComponent/gameBoardComponent.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    }
+  }
+
+
+  handleKeyDown = (event) => {
+    const keyNumber = event.keyCode;
+    const { key } = event;
+    // we only send the event if the event is right, left, top or bottom arrow
+    // these integers are numeric value of these arrows on keyboard.
+    if ([37, 38, 39, 40].find(number => number === keyNumber)) {
```
Comment (niinpatel) [inline client/src/components/gameBoardComponent/gameBoardComponent.js]: I'd say don't use keycodes. From the [documentation](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode):

> "This feature is no longer recommended. Though some browsers might still support it, it may have already been removed from the relevant web standards, may be in the process of being dropped, or may only be kept for compatibility purposes."

Comment (prakhar1912) [inline client/src/components/gameBoardComponent/gameBoardComponent.js]: Or
```js
[37,38,39,40].includes(keyNumber)
```

File: client/src/App.js [hunk 7]
```diff
@@ -1,8 +1,10 @@
 import React, { Component } from 'react';
+import { Route, BrowserRouter as Router } from 'react-router-dom';
 import { decode } from 'jsonwebtoken';
 
 import Navbar from './components/Navbar';
 import Home from './components/Home';
+import GameBoardComponent from './components/gameBoardComponent/gameBoardComponent';
```
Comment (niinpatel) [inline client/src/App.js]: Two things, 

1. File name for component should be capitalized. 
2. `components/gameBoardComponent/gameBoardComponent.js` is a bit redundant. you could either use 

`components/GameBoardComponent.js` or `components/GameBoardComponent/index.js`

File: client/src/components/gameBoardComponent/gameBoardComponent.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+      // if user comes directly to game board url, without request a game play,--
+      // -- will redirect to home page, where he/she/they can request to a new game play
+      return <Redirect to={{ pathname: '/' }} />;
+    }
+    return (
+      <div className="Board-container">
+        <img src={backgroundImage} id="backgroundImage" style={{ display: 'none' }} alt="backgroundImage" />
+        <canvas
+          ref={(canvas) => { this.canvas = canvas; }}
+          width="700"
```
Comment (niinpatel) [inline client/src/components/gameBoardComponent/gameBoardComponent.js]: move height and width to a constant file

File: client/src/components/Home.js (hunk 2 above)
Comment (niinpatel) [reply client/src/components/Home.js]: This should've been caught by the pre-commit hook, but I guess it didn't work. 

I'll add `yarn lint` in travis also so this kind of errors are caught before PR is submitted for review.

File: client/src/components/Home.js [hunk 9]
```diff
@@ -1,30 +1,98 @@
-import React from 'react';
+import React, { Component } from 'react';
+import { Redirect } from 'react-router-dom';
 import PropTypes from 'prop-types';
 import Button from '@material-ui/core/Button';
 import Icon from '@material-ui/core/Icon';
 import Grid from '@material-ui/core/Grid';
 import { withStyles } from '@material-ui/core/styles';
 import Leaderboard from './Leaderboard';
+import LoaderComponent from './common/loaderComponent/loaderComponent';
```
Comment (niinpatel) [inline client/src/components/Home.js]: capitalize component filename

File: client/src/components/common/loaderComponent/loaderComponent.css [hunk 10]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (niinpatel) [inline client/src/components/common/loaderComponent/loaderComponent.css]: I'd suggest use CSS-in-JS to be consistent with rest of the projects. See, `client\src\components\Navbar.js` for an example. But this is not a priority right now so we can make another PR for that.

Comment (vkbinfo) [reply client/src/components/common/loaderComponent/loaderComponent.css]: Let's have a discussion over it. I think a separate CSS file is a better approach.

File: client/src/components/common/loaderComponent/loaderComponent.js (hunk 5 above)
Comment (vkbinfo) [reply client/src/components/common/loaderComponent/loaderComponent.js]: Maafi. I will do that now.

File: client/src/components/Home.js (hunk 9 above)
Comment (vkbinfo) [reply client/src/components/Home.js]: Will refactor after Prakhar branch is merged because he has taken a branch out of this branch.

Comment (prakhar1912) [reply client/src/components/Home.js]: @vkbinfo You can go ahead and refactor. I can always rebase and resolve conflicts.

**Comment by niinpatel** [comment]
LGTM
